Python packaging for PyNEST-NG

steffengraber opened the pull request:
## Key changes:

- Python packaging 
  - Add Python packaging configuration with project metadata, dependencies, and optional features.

- CI Workflow  
  - Update the GitHub Actions workflow to improve wheel building and eployment to testpypi (https://test.pypi.org/project/nest-simulator/).

- CMake Files
  - Refactor CMake configuration to build the Python Wheel.
  
Before merging, the project name and deployment secrets need to be adjusted.

## The Python package may be tested in a Docker container:

```bash
docker run -it --rm --name test ubuntu:24.04 bash
```

### Install Dependencies and Create an Python venv

```bash 
apt update
DEBIAN_FRONTEND=noninteractive TZ=Etc/UTC apt install -y git python3 python3-pip python3-venv time libopenmpi-dev libboost-dev libgsl-dev
python3 -m venv ${HOME}/.env
source ${HOME}/.env/bin/activate
``` 

### Intalll from https://test.pypi.org/project/nest-simulator/

```bash 
python3 -m pip --no-cache-dir install --index-url https://test.pypi.org/simple/ --extra-index-url https://pypi.org/simple/ nest-simulator
python3 -c "import nest; nest.help()"
``` 
### Rebuild with MPI

``` bash
python3 -m pip install -v  --force-reinstall --check-build-dependencies --index-url https://test.pypi.org/simple/  --extra-index-url https://pypi.org/simple/  --config-settings=cmake.define.with-mpi="ON" --no-binary nest-simulator nest-simulator[mpi]
python3 -c "import nest; nest.help()"
``` 

### Install and Build from GitHub Repository

``` bash
python3 -m pip install -v --force-reinstall git+https://{GITHUB_TOKEN}@github.com/steffengraber/nest-simulator.git@hep_pynest-ng
python3 -c "import nest; nest.help()"
``` 

## Tested on:

- [x] Ubuntu 24.04 
- [x] Debian trixie
- [x] Alpine 3.23.0
- [x] WSL

steffengraber committed:
Add PEP 518 Python project setup for building pip packages

steffengraber committed:
Add cibuildwheel  worklflow to build and deploy pypi compilant pip packges

steffengraber committed:
Update pyproject.toml to hatchling builds and support for Python 3.13

steffengraber committed:
Update installation paths in CMakeLists.txt for nest and examples directories

steffengraber committed:
Update version to 3.8.post0.dev2 and adjust cibuildwheel build targets

steffengraber committed:
Update package installation command to use 'pynest-ng' instead of 'nest-simulator'

steffengraber committed:
Bump version to 3.8.post0.dev3 in pyproject.toml

steffengraber committed:
Remove NEST binary command check and update example selection logic in run_examples.sh

steffengraber committed:
Update CMake installation paths for SKBUILD compatibility and bump version to 3.8.post0.dev4

steffengraber committed:
Update Python version to 3.12 and modify test setup in build workflow

steffengraber committed:
Bump version to 3.8.post0.dev5 in pyproject.toml

steffengraber committed:
Merge remote-tracking branch 'heplesser/nest-simulator/pynest-ng' into hep_pynest-ng

heplesser commented on pyproject.toml:
```diff
@@ -1,3 +1,183 @@
+"Bug Tracker" = "https://github.com/nest/nest-simulator/issues"
+Changelog = "https://nest-simulator.readthedocs.io/en/stable/whats_new/index.html"
+
+[tool.scikit-build]
```
Why "scikit"?

steffengraber replied:
I use a program called "Scikit-build-core" to help me package pynest-ng. "It provides a bridge between CMake and the Python build system, allowing you to make Python modules with CMake." https://scikit-build-core.readthedocs.io/en/latest/
The scikit tag here marks some definitions for the build process.

heplesser commented on pyproject.toml:
```diff
@@ -1,3 +1,183 @@
+]
+dependencies = [
+  "numpy",
+  "pynest-api",
```
Maybe add pytest, so one can easily run the testsuite after installation?

steffengraber replied:
Pytest has been added as a dependency.

heplesser commented on pyproject.toml:
```diff
@@ -1,3 +1,183 @@
+  NEST_SRC_DIR=${PWD}
+  if [ $AUDITWHEEL_POLICY == "musllinux_1_2" ]
+  then
+    apk add git readline readline-dev libc6-compat gsl gsl-dev ncurses ncurses-dev wget bzip2 hdf5 cmake
```
For PyNEST-NG, we should no longer need `readline` and `readline-dev`.

steffengraber replied:
:+1:  Readline has been removed.

heplesser reviewed (commented):
Hi @steffengraber! Thanks for this PR and apologies for taking weeks before looking at it. I have tried the pip install and that worked very nicely on my macOS 15.5 (arm64) system. I could also run all non-MPI tests with 
```
pytest --ignore=$PyTestDir/mpi --ignore=$PyTestDir/sli2py_mpi $PyTestDir
```
where `PyTestDir` points to `.../testsuite/pytests` under the `site_packages/nest` directory.

Building with MPI did not quite work. NEST built, but the CMake output did not even show a search for OpenMPI and NEST then built nicely without MPI. One problem I had was that my zsh shell did not like the line
```
python3 -m pip install -v  --force-reinstall --check-build-dependencies --index-url https://test.pypi.org/simple/  --extra-index-url https://pypi.org/simple/  --config-settings=cmake.define.with-mpi="ON" --no-binary pynest-ng pynest-ng[mpi]
```
I had to put the final `pynest-ng[mpi]` in quotes to make this run. Could that have caused the problem?

What are the `SKBUILD` switches in the CMake files about? Probably a dumb question ...

You also mentioned that some secrets needed adjustment before merging. Do you mean merging this branch into my branch or the eventual merge into NEST master?

steffengraber commented:
> I had to put the final `pynest-ng[mpi]` in quotes to make this run. Could that have caused the problem?

Maybe a Mac thing?

> What are the `SKBUILD` switches in the CMake files about? Probably a dumb question ...

For Phyton packaging, some installation paths need to be adjusted. The SKBUILB switch recognizes that a build is being performed with Scikit-build-core and the corresponding changes can be made.
 
> You also mentioned that some secrets needed adjustment before merging. Do you mean merging this branch into my branch or the eventual merge into NEST master?

Both. In order for CI to perform the upload to testpypi or later to pypi, a token must be set as an environment variable.

steffengraber committed:
Add pytest to project dependencies

steffengraber committed:
Remove  readline installation

steffengraber committed:
Refactor verbosity setting in tests and update test structure

- Changed the method of setting verbosity in multiple test files from `nest.set_verbosity(nest.verbosity.M_WARNING)` to `nest.verbosity = nest.VerbosityLevel.WARNING` for consistency.
- Updated the `test_tsodyks2_synapse.py` to use pytest instead of unittest, including parameterization for `tau_fac`.
- Added checks for MPI support in `issue-1703.py` and improved error handling.
- Enhanced `run_test.sh` for better variable handling and clarity.
- Modified `summarize_tests.py` to include expected test counts and report missing tests.
- Cleaned up imports and ensured consistent formatting across utility files.

steffengraber committed:
Bump version to 3.8.post0.dev6 in pyproject.toml

steffengraber committed:
Add new tests and update existing ones for improved functionality and coverage

- Introduced new tests for Ginzburg neurons, global RNG, rate neurons, and sinusoidal generators to ensure correct behavior in parallel simulations.
- Updated existing tests to improve data handling, including CSV output formatting and connection validation for various neuron models.
- Added regression tests for specific tickets addressing issues with receptor types, spike generator behavior, and synapse connections.
- Enhanced the test suite summary to reflect the increased number of tests for SLI2Py MPI.

steffengraber committed:
Update project name and version in pyproject.toml

steffengraber committed:
Update project name to "pynest-ng" and version to "3.8.post0.dev7" in pyproject.toml

steffengraber committed:
Merge remote-tracking branch 'heplesser/nest-simulator/pynest-ng' into hep_pynest-ng

steffengraber committed:
Bump version to 3.8.post0.dev8 in pyproject.toml

steffengraber committed:
Fix command syntax for running tests in the CI workflow

steffengraber committed:
Update build-system requirements and adjust project versioning in pyproject.toml

steffengraber committed:
Bump version to 3.9.0-post0.dev9 in VERSION file

steffengraber committed:
Update macOS version in CI workflow to macos-15. macos-13 lable is deprecated.

steffengraber committed:
Update scikit-build metadata configuration and adjust brew install commands

steffengraber committed:
Update CI workflow to use macos-latest instead of macos-15

steffengraber committed:
Add regex for version extraction in scikit-build metadata

steffengraber committed:
Update CI workflow to use macos-15-intel for building wheels and add example script execution

steffengraber committed:
Merge remote-tracking branch 'heplesser/nest-simulator/pynest-ng' into hep_pynest-ng

steffengraber committed:
Update macOS version in CI workflow and project configuration to 15.0

steffengraber committed:
Fix brew install command formatting in macOS build script

steffengraber committed:
Refactor dependencies in pyproject.toml and enhance optional dependencies for desktop, server, test, and docs

steffengraber committed:
Add NEST pip installation and first steps guide

steffengraber committed:
Update included features

steffengraber committed:
Bump version to 3.9.0-post0.dev10 and add command-line interface entry point

steffengraber committed:
Merge remote-tracking branch 'heplesser/nest-simulator/pynest-ng' into hep_pynest-ng

steffengraber committed:
Fix formatting of SKBUILD condition checks in CMakeLists.txt files (pre-commit)

steffengraber committed:
Fix installation destination for scripts in CMakeLists.txt

steffengraber committed:
Fix after discussion

steffengraber committed:
Fix after discussion

steffengraber committed:
Add testing step for built packages in CI workflow and update version to 3.9.0-post0.dev11

steffengraber committed:
Refactor dependencies in pyproject.toml: remove unused packages and add to examples

steffengraber committed:
Merge remote-tracking branch 'heplesser/nest-simulator/pynest-ng' into hep_pynest-ng

steffengraber committed:
Update CI workflow to support additional OS versions and bump version to 3.9.0-post0.dev12

steffengraber committed:
Remove duplicate OS entries from CI workflow matrix

steffengraber committed:
Set minimal python version to 3.10

steffengraber committed:
Merge remote-tracking branch 'heplesser/nest-simulator/pynest-ng' into hep_pynest-ng

steffengraber committed:
Refactor NEST import messages and remove unused setup.py.in file

steffengraber committed:
Comment out test package workflow in build_nest_wheel.yml

steffengraber committed:
Update upload_pypi job dependencies to require build_wheels

steffengraber committed:
Update dependencies and remove Python 3.14 classifier from pyproject.toml

steffengraber committed:
Update GitHub Actions workflow and project metadata

- Upgrade actions/checkout and actions/upload-artifact to v6 in build_nest_wheel.yml
- Re-enable test_packages job with improved testing steps
- Bump version to 3.9.0-post0.dev13 in VERSION file
- Update project classifiers and dependencies in pyproject.toml

steffengraber committed:
Update GitHub Actions workflow to use consistent action versions

steffengraber committed:
Update GitHub Actions workflow: downgrade checkout action and adjust job dependencies

steffengraber committed:
Enhance GitHub Actions workflow to support full package variant and update project dependencies

steffengraber committed:
Update Cython dependency to remove upper version limit

steffengraber committed:
Update upload_pypi job dependencies to include build_wheels and build_sdist

steffengraber committed:
Enhance GitHub Actions workflow with wheel validation and retry logic for package installation

steffengraber committed:
Update macOS test-skip and deployment target in cibuildwheel configuration

steffengraber committed:
Refactor GitHub Actions workflow to simplify wheel and source distribution builds, removing unnecessary package type handling and validation steps.

steffengraber committed:
Update macOS deployment target to 15.0 in cibuildwheel configuration

steffengraber committed:
Refactor GitHub Actions workflow and update dependencies in pyproject.toml

steffengraber committed:
Refactor system dependency installation in before-all script for improved compatibility across package managers

steffengraber committed:
Comment out test command in cibuildwheel configuration for clarity

steffengraber committed:
Removed macos-14 builder

steffengraber committed:
Fix CMake configuration to enable HDF5 support instead of Sonata

steffengraber committed:
Disable HDF5 support in CMake configuration due to path issues

steffengraber committed:
Bump version to 3.9.0-post0.dev14

steffengraber committed:
Restore full Python version support in cibuildwheel configuration

steffengraber committed:
Merge remote-tracking branch 'heplesser/nest-simulator/pynest-ng' into hep_pynest-ng

steffengraber committed:
Merge remote-tracking branch 'heplesser/nest-simulator/pynest-ng' into hep_pynest-ng

steffengraber committed:
Bump version to 3.9.0-post0.dev15

steffengraber committed:
Enable HDF5 support in build configuration

steffengraber committed:
Add test dependencies and enhance test command in pyproject.toml

steffengraber committed:
Bump version to 3.9.0-post0.dev16

steffengraber committed:
Enable Sonata support and update HDF5 installation instructions in pyproject.toml

steffengraber committed:
Refactor HDF5 and Sonata support in pyproject.toml; update dependencies and build configuration

steffengraber committed:
Add cibuildwheel overrides for ARM and manylinux builds; adjust test requirements

steffengraber committed:
Update pip installation guide: refine prerequisites, clarify package name, and enhance optional features section

steffengraber committed:
Rename project to "nest-simulator" and update optional dependencies for Nest server support

steffengraber committed:
Update package name references from "pynest-ng" to "nest-simulator" in installation guide and workflow; revert version to 3.9.0-post0.dev0

steffengraber committed:
Merge remote-tracking branch 'heplesser/nest-simulator/pynest-ng' into hep_pynest-ng

steffengraber committed:
Update GitHub Actions workflow to trigger on release events and workflow dispatch

steffengraber committed:
Refactor GitHub Actions workflow: remove unused config-file parameter and simplify condition for upload step

steffengraber committed:
Enhance GitHub Actions workflow: add triggers for pull requests and pushes, expand testing on additional Linux distributions, and improve NEST functionality tests

steffengraber committed:
Update GitHub Actions workflow: remove Ubuntu 20.04 from build matrix and add it to testing on different Linux distributions

steffengraber committed:
Update pyproject.toml: disable OpenMP support and streamline macOS build dependencies

steffengraber committed:
Update pyproject.toml: clarify OpenMP support status for macOS compatibility

steffengraber committed:
Enhance GitHub Actions workflow: add macOS-specific testing and expand Linux distribution testing matrix

steffengraber committed:
Remove NEST functionality tests from GitHub Actions workflow

steffengraber committed:
Remove test_uploaded_package job from GitHub Actions workflow

steffengraber committed:
Update GitHub Actions workflow: refine macOS testing matrix and remove outdated Ubuntu versions

steffengraber committed:
Enhance GitHub Actions workflow: add local wheel testing and improve TestPyPI functionality tests

steffengraber committed:
Bump version to 3.9.0-post0.dev1

heplesser commented on pyproject.toml:
```diff
@@ -138,7 +138,7 @@ with-warning = "ON"
 with-boost = "ON"
 with-gsl = "ON"
-with-openmp = "ON"
+with-openmp = "OFF"
```
@steffengraber Here we need to discuss more and find a solution that allows us to use OpenMP. Without threads, NEST is not very attractive given that most laptops these days have at least for cores.

steffengraber replied:
I am currently working on a solution that is a little more flexible. I am also working on CI tests that should detect such problems under macOS as well.
I have set an “off” here to at least get a package that also works under macOS.
This package should be available within the next hour.

steffengraber committed:
Improve wheel file detection in local testing step of GitHub Actions workflow

steffengraber committed:
Update Python version classifiers and enhance macOS build setup in pyproject.toml

steffengraber committed:
Enhance macOS dependency installation with error handling and verification steps

steffengraber committed:
Enhance macOS dependency installation with architecture-specific handling and verification

steffengraber committed:
Refactor macOS setup script for unified architecture handling and dynamic path detection

steffengraber committed:
Update macOS deployment target configuration for improved compatibility

steffengraber committed:
Refactor macOS setup for ARM64 support and simplify dependency installation

steffengraber committed:
Enhance macOS build configuration for OpenMP with improved flags and paths

steffengraber committed:
Fix macOS ARM64 setup script to ensure proper export of CMAKE_ARGS and improve completion message

steffengraber committed:
Enhance macOS ARM64 setup script by adding OpenMP library paths and suffixes for improved compatibility

steffengraber committed:
Refactor macOS ARM64 setup script to streamline dependency installation and enhance OpenMP configuration

steffengraber committed:
Update macOS deployment target to 15.0 for compatibility with Homebrew libraries

steffengraber committed:
Refactor macOS ARM64 setup script to improve architecture verification and streamline dependency installation

steffengraber committed:
Enhance CI workflow by adding pip package caching and support for multiple macOS architectures; update version to 3.9.0-post0.dev1

jessica-mitchell committed:
Modify text and structure of install page (#43)

* update text

* modify text

steffengraber committed:
Update macOS job configuration to use macos-15-intel for improved compatibility

steffengraber committed:
Improve system dependency installation in CI by adding conditional sudo support for package managers

steffengraber committed:
Bump version to 3.9.0-post0.dev2 for release

steffengraber committed:
Enhance GitHub Actions workflow: add pip caching, update macOS setup, and enable OpenMP support

steffengraber committed:
Update installation instructions for macOS to include OpenMP requirements and bump version to 3.9.0-post0.dev3

steffengraber committed:
Refactor GitHub Actions workflow: simplify triggers, remove unused jobs, and specify config file for wheel building

steffengraber committed:
Refactor GitHub Actions workflow: enhance job configurations, add local wheel testing, and improve matrix strategy

steffengraber committed:
Fix TODO comments formatting in GitHub Actions workflow

steffengraber committed:
Enhance local wheel testing in GitHub Actions: add Python version compatibility checks and install missing dependencies; bump version to 3.9.0-post0.dev4

steffengraber commented:
In macOS, the brew path is now automatically detected and set. 
For macOS users, OpenMP support requires installing the OpenMP runtime library:
```
brew install libomp
pip install nest-simulator
```
@heplesser Could you test it again now?

I also have parallel testing of Ubuntu and macOS ARM packages and after uploading, there are now tests for Ubuntu, Debian, Alpine, CentOS, Fedora, and macOS Intel/ARM.
The tests are still a bit of a work in progress.

steffengraber committed:
Enhance local wheel testing: improve wheel selection logic, add debugging output, and ensure compatibility with manylinux; bump version to 3.9.0-post0.dev5

steffengraber committed:
Enhance local wheel testing: improve wheel selection logic with OS and architecture checks, add virtual environment setup, and streamline installation process

steffengraber committed:
Update container images in CI workflow: replace CentOS 7 with Rocky Linux 8 for improved compatibility

steffengraber committed:
Enhance Alpine compatibility in CI workflow: add gcompat for glibc support and implement error handling for package installation on Alpine Linux

steffengraber committed:
Update upload dependencies in CI workflow: change dependency order for uploading wheels to PyPI and bump version to 3.9.0-post0.dev6

steffengraber committed:
Fix virtual environment activation and improve Alpine package installation handling

steffengraber committed:
Bump version to 3.9.0-post0.dev9

steffengraber committed:
Update Python version to 3.10 because of issues in Rocky Linux  and enhance Alpine package installation in CI workflow

steffengraber committed:
Bump version to 3.9.0-post0.dev10

steffengraber committed:
Update Python version to 3.12 and remove Rocky Linux container from CI workflow; bump version to 3.9.0-post0.dev11, fixes a bug where $SUDO_CMDapk was missing a space

steffengraber committed:
Remove Alpine container from CI workflow and bump version to 3.9.0-post0.dev12

steffengraber committed:
Update Homebrew installation commands to install Boost separately and ensure proper linking

steffengraber committed:
Update Homebrew installation commands to include specific Boost version and streamline linking process

steffengraber committed:
Update Homebrew command to force link libomp for better dependency management

steffengraber committed:
Update Homebrew installation commands to use Boost 1.90.0 and streamline dependency management

steffengraber committed:
Update Homebrew installation commands to use Boost 1.88 and ensure proper linking

steffengraber committed:
Update Homebrew installation commands to use Boost 1.85 for compatibility

steffengraber committed:
Merge remote-tracking branch 'heplesser/nest-simulator/pynest-ng' into hep_pynest-ng

steffengraber committed:
Bump version to 3.9.0-post0.dev13

steffengraber committed:
Enhance build configuration by adding OpenBLAS installation and adjusting pip settings for binary wheel preference

steffengraber committed:
Update test dependencies and build configuration for improved compatibility

gtrensch commented:
I tested the three installation options with clean Ubuntu 22.04 and 24.04 installations. I also ran some of the NEST examples for quick testing (with and without MPI). Everything worked without any problems! :+1:

heplesser reviewed (approved):
@steffengraber Looks good, thanks!

gtrensch reviewed (approved):
Looks good to me, thanks! 👍

steffengraber committed:
Update build configuration to include macos-26 and correct branch name for pushes

steffengraber committed:
Remove low-level API version setting from NestModule class

steffengraber committed:
Bump version to 3.9.0-post0.dev14 and update Python 3.14 classifier in pyproject.toml

steffengraber committed:
Refactor NestModule attributes for improved readability and consistency with a new black version

steffengraber committed:
Update branch triggers in workflow and adjust macOS deployment target for compatibility

steffengraber committed:
Update NEST version retrieval methods in tests and documentation

steffengraber committed:
Update local wheel test to use nest.help() for better output

steffengraber committed:
Update macOS setup script for improved architecture detection and compatibility

steffengraber committed:
Update macOS build configurations for improved compatibility

steffengraber committed:
Refine conditions for uploading wheels to PyPI to ensure it only runs on releases or manual dispatch from non-fork branches
